### PR TITLE
Fix RequestBody logging when zero byte read is used

### LINKED ISSUE #47216: Different logging behaviour between net 6 and 7 for Request Body in HttpLoggingMiddleware
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If you allow the http-logger, which is running unter net 6 to log requests, it will log them as expected. But the same code unter net 7 shows just an empty value.

In .net 6
Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware: Information: RequestBody: Date=2023-03-15

In .net 7
Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware: Information: RequestBody:

My guess is that the https://github.com/dotnet/dotnet/blob/b5fc13cdacdec01bb18d3271d8606e96949943ad/src/runtime/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs#L243 here is the issue.

If used with .net 6 the value of the property reader.UseZeroByteReads is false, but its true if using .net 7
this invokes https://github.com/dotnet/aspnetcore/blob/bec278eabea54f63da15e10e654bdfa4168a2479/src/Middleware/HttpLogging/src/RequestBufferingStream.cs#L89 with an empty string and prevet logging the true value.

### Expected Behavior

In Log should be something like this
Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware: Information: RequestBody: Date=2023-03-15

### Steps To Reproduce

Here is the complete code: https://github.com/EnemyArea/BrokenAddHttpLoggingNet

**Program.cs**
```c#
builder.Services.AddHttpLogging(logging =>
{
    logging.LoggingFields = HttpLoggingFields.All;
    logging.MediaTypeOptions.AddText("application/x-www-form-urlencoded");
});

var app = builder.Build();
app.UseHttpLogging();
app.UseHttpsRedirection();
app.UseAuthorization();
app.MapControllers();
app.Run();
```

**WeatherForecastController.cs**
```c#
[Route("bydate")]
[HttpPost]
public IEnumerable<WeatherForecast> GetWeatherForecastForDate([FromForm] WeatherForecastRequest request)
{
    return Enumerable.Range(1, 5).Select(index => new WeatherForecast
    {
        Date = request.Date.AddDays(index),
        TemperatureC = Random.Shared.Next(-20, 55),
        Summary = Summaries[Random.Shared.Next(Summaries.Length)]
    })
    .ToArray();
}

appsettings.Development.json
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft.AspNetCore.HttpLogging.HttpLoggingMiddleware": "Information"
    }
  }
}
```

### Exceptions (if any)

_No response_

### .NET Version

7.0.101

### Anything else?
```
.NET SDK:
 Version:   7.0.101
 Commit:    bb24aafa11

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.101\

Host:
  Version:      7.0.1
  Architecture: x64
  Commit:       97203d38ba

.NET SDKs installed:
  3.1.426 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.203 [C:\Program Files\dotnet\sdk]
  6.0.309 [C:\Program Files\dotnet\sdk]
  7.0.101 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/47216

### CODE REVIEW

File: src/Middleware/HttpLogging/test/HttpLoggingMiddlewareTests.cs [hunk 1]
```diff
@@ -434,6 +434,33 @@ public async Task PartialReadBodyStillLogs()
         Assert.Contains(TestSink.Writes, w => w.Message.Equals("RequestBody: " + expected));
     }
 
+    [Fact]
+    public async Task ZeroByteReadStillLogsRequestBody()
+    {
+        var input = string.Concat(new string('a', 60000), new string('b', 3000));
```
Comment (amcasey) [inline src/Middleware/HttpLogging/test/HttpLoggingMiddlewareTests.cs:440]: Are the `b`s important?

File: src/Middleware/HttpLogging/src/RequestBufferingStream.cs [hunk 2]
```diff
@@ -27,6 +27,12 @@ public override async ValueTask<int> ReadAsync(Memory<byte> destination, Cancell
     {
         var res = await _innerStream.ReadAsync(destination, cancellationToken);
 
+        // Zero-byte reads can occur when using PipeReader, we don't want to accidentally complete the RequestBody logging in this case
```
Comment (amcasey) [inline src/Middleware/HttpLogging/src/RequestBufferingStream.cs:30]: From context, I assume passing zero bytes to `WriteToBuffer` would "complete the RequestBody logging"?

**Review by amcasey — APPROVED** [review]
Minor questions, but seems right

File: src/Middleware/HttpLogging/src/RequestBufferingStream.cs [hunk 3]
```diff
@@ -27,6 +27,12 @@ public override async ValueTask<int> ReadAsync(Memory<byte> destination, Cancell
...
+        if (destination.IsEmpty)
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: What about the one on [line 60](https://github.com/dotnet/aspnetcore/pull/47794/files#diff-9d3086eac5e1da5350c127d51efbb6f16fa3134e4337fbed15e3a07bb8f2b940R60)?

**Review by amcasey — COMMENTED** [review]
Actually, I'm confused - how does `WriteToBuffer` know when to `LogRequestBody`?

File: src/Middleware/HttpLogging/src/RequestBufferingStream.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: If we checked `res` instead we'd still need to check if `destination.IsEmpty`, because you could reach the end of the underlying stream which would give `res == 0`. So it's simpler just to check `destination.IsEmpty`.

File: src/Middleware/HttpLogging/test/HttpLoggingMiddlewareTests.cs (hunk 1 above)
Comment (BrennanConroy) [reply src/Middleware/HttpLogging/test/HttpLoggingMiddlewareTests.cs:440]: Idk, copied from the test above

File: src/Middleware/HttpLogging/src/RequestBufferingStream.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: I guess the one on line 60 should be updated for consistency. The reason I originally didn't change it is because zero-byte reads are almost always done via PipeReader, which doesn't have a synchronous read so Read won't be called. And, by default ASP.NET Core doesn't allow synchronous reads on streams anyways.

But, if a user set the option to allow synchronous reads, and did `request.Body.Read(new byte[0])`, then they would hit this path. It would be really odd, since the whole point of zero-byte reads is to avoid allocating until the underlying stream has data, which would be an async operation if there was no data yet.

TLDR; I'll update it for consistency 😄

[the remark this answers is not in the document]
Comment (amcasey) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: I was not reading this correctly.  "zero-byte reads" refers to the case where the destination buffer is empty, rather than to the case where the stream writes zero bytes to the buffer.

**Review by JamesNK — APPROVED** [review]
Zero byte reads strike again

File: src/Middleware/HttpLogging/src/RequestBufferingStream.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: Zero-byte reads allow code to wait for available data without providing a potentially big buffer while waiting.

For example:
1. Make zero byte read to wait for data to be available: `await stream.ReadAsync(Array.Empy<byte>())`
2. Zero-byte read returns. Rent buffer from pool and read again with a buffer: `await stream.ReadAsync(oneMegabyteBuffer)`. It should return quickly, allowing the buffer to be returned to the pool quickly.

[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: > TLDR; I'll update it for consistency 😄

Are zero-byte reads for sync `Stream.Read` a thing? I haven't use sync read methods in so long that I'm not sure.

**Review by captainsafia — APPROVED** [review]
LGTM! I don't think we need to update L60 as mentioned in https://github.com/dotnet/aspnetcore/pull/47794/files#r1171970619

File: src/Middleware/HttpLogging/src/RequestBufferingStream.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: I'm fine with asserting it never happens if that's the case.

[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: > Are zero-byte reads for sync Stream.Read a thing?

I think they are, because lots of stream implementations just call the async api and block, so they inherit the zero-byte read feature from the async code path.

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Middleware/HttpLogging/src/RequestBufferingStream.cs:31]: Yes, be consistent for Read and ReadAsync.

**Review by Tratcher — APPROVED** [review]
